ens-improvement-proposals/ensip-16-offchain-metadata.md

```diff
@@ -0,0 +1,168 @@
+
+#### Dynamic Metadata
+
+Metadata serves a crucial role in providing valuable insights about a node owner and their specific resolver. In certain scenarios, resolvers may choose to adopt diverse approaches to resolve data based on the node. An example of this would be handling subdomains of a particular node differently. For instance, we could resolve "optimism.foo.eth" using a contract on optimism and "gnosis.foo.eth" using a contract on gnosis.
+By passing the name through metadata, we empower the resolution process, enabling CcipResolve flows to become remarkably flexible and scalable. This level of adaptability ensures that our system can accommodate a wide array of use cases, making it more user-friendly and accommodating for a diverse range of scenarios.
+
+### Implementation
+
+```solidity
+
+interface OffChainResolver {
+    /** @dev Returns the owner of the resolver on L2
+     * @param node
+     * @return owner in bytes32 instead of address to cater for non EVM based owner information
+     */
+    owner(bytes32 node) returns (bytes32 owner);
+
+    isApprovedForAll(address account, address operator) returns (boolean);
+    /** @dev Returns the owner of the resolver on L2
+     * @return name can be l2 chain name or url if offchain
+     * @return coinType according to https://github.com/ensdomains/address-encoder
+     * @return graphqlUrl url of graphql endpoint that provides additional information about the offchain name and its subdomains
+     * @return storageType 0 = EVM, 1 = Non blockchain, 2 = Starknet
+     * @storageLocation = l2 contract address
+     * @return context = an arbitrary bytes string to define the namespace to which a record belongs such as the name owner.
+     */
+    function metadata(bytes calldata name)
+        external
+        view
+        returns (string memory, uint256, string memory, uint8, bytes memory, bytes memory)
+    {
+        return (name, coinType, graphqlUrl, storageType, storageLocation, context);
+    }
+
+    // Optional. If context is dynamic, the event won't be emitted.
+    event MetadataChanged(
+        string name,
+        uint256 coinType,
+        string graphqlUrl,
+        uint8 storageType,
+        bytes storageLocation,
+        bytes context
+    );
+}
+```
+
+```javascript
+const node = namehash('ccipreadsub.example.eth')
+const resolver = await ens.resolver(node)
+const owner = await resolver.owner(node)
+// 0x123...
+const dataLocation = await.resolver.graphqlUrl()
+// {
+//   url: 'http://example.com/ens/graphql',
+// }
+```
+
+##### GraphQL schema
```

Comment:
This needs a short description as to why it's being specified and how it will operate.

Reply:
I had in mind something more specific to this EIP, like "this specifies the schema for a standardized metadata subgraph, which collects information emitted by L1 and L2 resolvers".